astropy/units/typing.py

```diff
@@ -76,3 +76,4 @@
 Complex: TypeAlias = Real | complex | np.complexfloating
 
 UnitPower: TypeAlias = int | float | Fraction
+UnitScale: TypeAlias = int | float | Fraction | complex
```

Comment:
I'm a bit uncomfortable with adding the `complex` to this type annotation.
What are the pros / cons to adding a new type alias that extends the one w/out `complex` ?

Reply:
I'm confused: the unit scale can be complex, so why should the type not indicate that?

Reply:
It looks like it would be possible to define
```python
UnitScale: TypeAlias = UnitPower | complex
```
However, there is no meaningful conceptual relation between `UnitPower` and `UnitScale` that would justify using `UnitPower` in the `UnitScale` definition. Both of them are unions of numerical types so they end up having a large overlap, but that overlap is essentially a coincidence.

Reply:
I meant more that isn't the unit scale normally real, and the complex cases are rare. So there could be two type hints for the unit scale

Reply:
It shouldn't matter how common or how rare some type might be. What matters is if a given type is allowed or not, and `complex` is an allowed type for the scale of a unit.

Reply:
For now, sure. It's just easier to expand an API than restrict it. So if complex scalings are niche, perhaps it's best to keep them niche, even at the annotation level.

Reply:
They're definitely niche, but how would one go about hinting that?

Reply:
Type checkers don't care about how common or how rare some parameter types are, they only care about whether some given type is allowed or not. That is a good thing because that is what allows type checkers to discover weird edge cases that unit tests are not likely to stumble upon.

Reply:
This isn't a blocker to merging.
This all comes down to my initial reaction that it's surprising that there's nowhere in the code that `UnitScale` *must* be real.
Then we would want
```
RealUnitScale: TypeAlias = int | float | Fraction
UnitScale: TypeAlias = RealUnitScale | complex
```
in which case `RealUnitScale` would pass mypy everywhere `UnitScale` is used, and is more type narrow.
But if `UnitScale` can truly be complex everywhere, we should resolve and merge this.

Reply:
FWIW, complex scales were introduced to be able to raise units to arbitrary power even if they have negative scale. E.g., `u.mag` is just `-0.4*u.dex` and allowing complex values means it is possible to do,
```
In [5]: (u.mag.decompose()**0.5).scale
Out[5]: 0.6324555320336759j
```
See https://github.com/astropy/astropy/pull/1628#issuecomment-30388097

Perhaps this could have been solved differently, or perhaps it might have been better to just forbid it... Anyway, my 2¢ would be not to worry here and get this merged...

Reply:
p.s. I still like magnitudes, even though they clearly are one of the most troublesome units ever...

Reply:
> Anyway, my 2¢ would be not to worry here and get this merged. 

Sure :). Worth revisiting.